pkg/compute/store/boltdb/store.go

```diff
@@ -404,35 +444,64 @@ func (s *Store) deleteExecution(tx *bolt.Tx, executionID string) error {
 	}
 
 	// Delete the bucket with the execution-id within the history bucket
-	historyBucket := tx.Bucket([]byte(BucketHistoryName))
+	historyBucket := s.getHistoryBucket(tx)
 	err = historyBucket.DeleteBucket([]byte(executionID))
 	if err != nil {
-		errorList = errors.Join(errorList, err)
+		return err
 	}
 
-	return errorList
+	return nil
 }
 
 // Close ensures the database is closed cleanly
 func (s *Store) Close(ctx context.Context) error {
 	return s.database.Close()
 }
 
-func (s *Store) GetExecutionCount(ctx context.Context) (uint, error) {
+func (s *Store) GetExecutionCount(ctx context.Context, state store.ExecutionState) (uint, error) {
 	log.Ctx(ctx).Trace().
 		Msg("boltdb.GetExecutionCount")
 
-	// TODO(ross) So, counting based on execution state .... seems like
-	// maintaining a prefix for a counter is excessive
-
-	// var counter uint
-	// for _, execution := range s.executionMap {
-	// 	if execution.State == store.ExecutionStateCompleted {
-	// 		counter++
-	// 	}
-	// }
-	// return counter, nil
-	return 0, nil
+	// We have to wait here to ensure the counter has been populated,
+	// so we will wait until we know it has finished
+	s.starting.Wait()
+	return uint(s.stateCounter.Get(state)), nil
+}
+
+func (s *Store) populateStateCounter(ctx context.Context) {
+	acc := NewStateCounter()
+
+	err := s.database.View(func(tx *bolt.Tx) (err error) {
+		bucket := s.getExecutionsBucket(tx)
+		cursor := bucket.Cursor()
+
+		for k, v := cursor.First(); k != nil; k, v = cursor.Next() {
+			if v == nil { // We can't do much with empty values
+				continue
+			}
+
+			var entry store.Execution
+			err = json.Unmarshal(v, &entry)
+			if err != nil {
+				return
+			}
+
+			acc.IncrementState(entry.State, 1)
+		}
+
+		return err
+	})
+
+	if err != nil {
+		log.Ctx(ctx).Error().Err(err).Msg("failed to generate state counter for execution store")
+	}
+
+	// As more items may have been added whilst we were running, merge the
+	// statecounter we have generated into the existing one for the store
+	s.stateCounter.Include(acc)
```

Comment:
Is that possible given we are populating within a transaction? Does boltdb allow concurrent transactions on the same bucket?

Reply:
One transaction per goroutine, but otherwise the only other restriction seems to be that there's only a single writer txn at a time.